pending-checkmarks-cf-worker/DESIGN.md

```diff
@@ -0,0 +1,71 @@
+		4. if checkmark does not exist:
+			1. assign a checkmark to the wallet, retrieving `wallet_address` from the KV store using `pending_session_id`, with `checkmark_id` set to the sha-512 hash of `pending_session_id`
+			2. map `initial_session_id` to `pending_session_id`, replacing `current_session_id`, to indicate that this is the new session for which the checkmark is currently assigned associated with that first initial verified session
+			3. delete both `wallet_address` <=> `pending_session_id` mappings as they are no longer needed
+	- on other failure:
+		1. do nothing so wallet can view failure reason
+	- on success (must be first time verifying):
+		1. retrieve `wallet_address` from KV store using `pending_session_id`
+		2. assign checkmark to `wallet_address` with `checkmark_id` set to the sha-512 hash of `pending_session_id`
+		3. map `pending_session_id` to `pending_session_id` to indicate that this is the session for which the checkmark is currently assigned. the `initial_session_id` and `current_session_id` are equivalent, since this is the initial successful session and a checkmark is currently assigned for it
+		4. delete both `wallet_address` <=> `pending_session_id` mappings as they are no longer needed
+5. person tries to verify with a different wallet
+	1. get unique session ID from synaps (`pending_session_id`), and store two KV store keys: (1) mapping `pending_session_id` and `wallet_address`, and (2) mapping `wallet_address` to `pending_session_id`
+		- check if this rejects immediately or after some delay (i.e. will it instant-reject or will the webhook have to handle failures)
+	2. show pending status
+6. person deletes the checkmark, and tries to verify again, from any wallet
+	1. get unique session ID from synaps (`pending_session_id`), and store two KV store keys: (1) mapping `pending_session_id` and `wallet_address`, and (2) mapping `wallet_address` to `pending_session_id`
+		- check if this rejects immediately or after some delay (i.e. will it instant-reject or will the webhook have to handle uniqueness failures that turn into new checkmark assignments)
+	2. show pending status
+
+### Security concerns
+
+sensitive information to minimize exposure to:
+- personally identifiable information (PII)
+- connection between PII and a wallet address
+- connection between different wallets associated with the same PII, even if the PII is unknown
+
+the link between the `session_id` and `wallet_address` is only meaningful if you've compromised synaps or our API key because you can then retrieve session info containing PII. this link can only be discovered in the following situations:
+- session is pending or failed, and the attacker has access to the wallet to perform wallet signature auth
+- session is pending or failed, and the attacker has access to the KV store
```

Comment:
Great observations. Speaking of, when do the failed sessions get cleared? If they persist forever, we may have privacy issues if someone ever gets their wallet compromised..

Reply:
this is now slightly incorrect because we never send the session ID or PII back to the user after initial verification. all we show is their status. and the checkmark IDs are the hashes of session IDs. though we do persist failed sessions forever, so you'd have to compromise the cloudflare worker to find the link, and then synaps or our API key to get access to the data.